server/src/test/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponseTests.java

```diff
@@ -0,0 +1,112 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.indices.validate.query;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.support.DefaultShardOperationFailedException;
+import org.elasticsearch.action.support.broadcast.AbstractBroadcastResponseTestCase;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.test.AbstractStreamableXContentTestCase;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ValidateQueryResponseTests extends AbstractBroadcastResponseTestCase<ValidateQueryResponse> {
+
+    protected static ValidateQueryResponse createRandomValidateQueryResponse(
+        int totalShards, int successfulShards, int failedShards, List<DefaultShardOperationFailedException> failures) {
+        boolean valid = failedShards == 0;
+        List<QueryExplanation> queryExplanations = new ArrayList<>(totalShards);
+        for(DefaultShardOperationFailedException failure: failures) {
+            queryExplanations.add(
+                new QueryExplanation(
+                    failure.index(), failure.shardId(), false, failure.reason(), null
+                )
+            );
+        }
+        return new ValidateQueryResponse(
+            valid, queryExplanations, totalShards, successfulShards, failedShards, failures
+        );
+    }
+
+    protected static ValidateQueryResponse createRandomValidateQueryResponse() {
+        int totalShards = randomIntBetween(1, 10);
+        int successfulShards = randomIntBetween(0, totalShards);
+        int failedShards = totalShards - successfulShards;
+        boolean valid = failedShards == 0;
+        List<QueryExplanation> queryExplanations = new ArrayList<>(totalShards);
+        List<DefaultShardOperationFailedException> shardFailures = new ArrayList<>(failedShards);
+        for (int i=0; i<successfulShards; i++) {
+            QueryExplanation queryExplanation = QueryExplanationTests.createRandomQueryExplanation(true);
+            queryExplanations.add(queryExplanation);
+        }
+        for (int i=0; i<failedShards; i++) {
+            QueryExplanation queryExplanation = QueryExplanationTests.createRandomQueryExplanation(false);
+            ElasticsearchException exc = new ElasticsearchException("some_error_" + randomInt());
+            shardFailures.add(
+                new DefaultShardOperationFailedException(
+                    queryExplanation.getIndex(), queryExplanation.getShard(),
+                    exc
+                )
+            );
+            queryExplanations.add(queryExplanation);
+        }
+        Collections.shuffle(queryExplanations, random());
+        Collections.shuffle(shardFailures, random());
```

Comment:
why do we need to shuffle these two?

Reply:
We are creating the XContent thing very serially. First all successful ones and then all failures. I thought it would be more realistic if this was shuffled.

Reply:
but they go to two different lists, so I think that what matters is that each item is randomized? I am not sure but shuffling buys us.

Reply:
The query explanation contains the failures and the successful ones. Ok, we can do away with the shuffling of failures. But the query explanations should stay I think.